File: app/src/main/java/ru/ibakaidov/distypepro/components/InputGroup.java
package ru.ibakaidov.distypepro.components;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Spinner;

import androidx.annotation.Nullable;

import com.google.firebase.analytics.FirebaseAnalytics;

import ru.ibakaidov.distypepro.R;
import ru.ibakaidov.distypepro.screens.SpotlightActivity;
import ru.ibakaidov.distypepro.utils.Callback;
import ru.ibakaidov.distypepro.utils.ProgressState;
import ru.ibakaidov.distypepro.utils.TTS;

public class InputGroup extends Component {
    private TTS tts;
    private EditText ttsEditText;
    private Button sayButton;
    private ImageButton spotlightButton;
    private Spinner chatSpinner;
    private int prevSpinnerValue = 0;
    private String[] textCache = new String[] {"", "", ""};

    public InputGroup(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public void setTts(TTS tts) {
        this.tts = tts;

        tts.setOnPlayCallback(new Callback<ProgressState>() {
            @Override
            public void onDone(ProgressState res) {
                sayButton.setText(res == ProgressState.START?R.string.stop:R.string.say);
            }
        });
    }

    @Override
    protected void initUI() {
        ttsEditText = findViewById(R.id.text_to_speech_edittext);
        sayButton = findViewById(R.id.say_button);
        spotlightButton = findViewById(R.id.spotlight_button);


        setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                if(ttsEditText.hasFocus()){
                    ttsEditText.clearFocus();
                }
            }
        });
        View v = this;
        ttsEditText.setOnFocusChangeListener(new OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean b) {
                ViewGroup.LayoutParams params = v.getLayoutParams();
                params.height = b?LayoutParams.MATCH_PARENT: ViewGroup.LayoutParams.WRAP_CONTENT;
                v.setLayoutParams(params);
            }
        });

        sayButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                say();
            }
        });

        spotlightButton.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                spotlight();
            }
        });

    }

    private void setText(String s) {
        ttsEditText.setText(s);
    }

    private void spotlight() {
        SpotlightActivity
                .show(getContext(), getText());

        FirebaseAnalytics
                .getInstance(getContext())
                .logEvent("spotlight", null);
    }

    @Override
    protected int getLayoutId() {
        return R.layout.input_group;
    }

    private void say() {
        String text = getText();
        tts.speak(text);
        FirebaseAnalytics
                .getInstance(getContext())
                .logEvent("say", null);
    }

    private String getText() {
        return ttsEditText.getText().toString();
    }

    public void setChatSpinner(Spinner chatSpinner) {
        this.chatSpinner = chatSpinner;
        ArrayAdapter adapter = new ArrayAdapter<String >(getContext(), androidx.appcompat.R.layout.support_simple_spinner_dropdown_item, new String[]{"1", "2", "3"});
        chatSpinner.setAdapter(adapter);

        chatSpinner
                .setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
                    @Override
                    public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                        textCache[prevSpinnerValue] = getText();
                        setText(textCache[i]);
                        prevSpinnerValue = i;

                    }

                    @Override
                    public void onNothingSelected(AdapterView<?> adapterView) {

                    }
                });
    }

    public void clear() {
        ttsEditText.setText("");
    }

    public void back() {
        ttsEditText.clearFocus();
    }
}

File: app/src/main/java/ru/ibakaidov/distypepro/components/InputGroup.kt
package ru.ibakaidov.distypepro.components

import android.content.Context
import android.util.AttributeSet
import android.widget.EditText
import android.widget.Toast
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.findViewTreeLifecycleOwner
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.google.android.material.button.MaterialButton
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import ru.ibakaidov.distypepro.R
import ru.ibakaidov.distypepro.screens.SpotlightActivity
import ru.ibakaidov.distypepro.utils.Callback
import ru.ibakaidov.distypepro.utils.ProgressState
import ru.ibakaidov.distypepro.utils.Tts
import ru.ibakaidov.distypepro.utils.Tts.TtsEvent

class InputGroup @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null
) : Component(context, attrs) {

    override fun layoutId(): Int = R.layout.input_group

    private lateinit var tts: Tts
    private lateinit var ttsEditText: EditText
    private lateinit var sayButton: MaterialButton
    private val textCache = arrayOf("", "", "")
    private var eventsJob: Job? = null
    private var isSpeaking: Boolean = false

    override fun initUi() {
        ttsEditText = findViewById(R.id.text_to_speech_edittext)
        sayButton = findViewById(R.id.say_button)

        setOnClickListener {
            if (ttsEditText.hasFocus()) {
                ttsEditText.clearFocus()
            }
        }

        sayButton.setOnClickListener { say() }

        ttsEditText.addTextChangedListener(object : SimpleTextWatcher() {
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                updateActionButtonsState()
            }
        })

        updateActionButtonsState()
    }

    fun setTts(tts: Tts) {
        this.tts = tts
        tts.setOnPlayCallback(object : Callback<ProgressState> {
            override fun onDone(result: ProgressState) {
                when (result) {
                    ProgressState.START -> {
                        isSpeaking = true
                        sayButton.setText(R.string.stop)
                        sayButton.isEnabled = true
                    }
                    ProgressState.STOP, ProgressState.ERROR -> {
                        isSpeaking = false
                        sayButton.setText(R.string.say)
                        updateActionButtonsState()
                    }
                }
            }
        })
        observeTtsEvents()
    }

    fun switchSlot(fromSlot: Int, toSlot: Int) {
        textCache[fromSlot] = currentText
        setText(textCache[toSlot])
    }

    fun clear() {
        ttsEditText.setText("")
    }

    fun back() {
        ttsEditText.clearFocus()
    }

    private fun say() {
        val text = currentText
        tts.speak(text)
        Firebase.analytics.logEvent("say", null)
    }


    fun spotlight() {
        SpotlightActivity.show(context, currentText)
        Firebase.analytics.logEvent("spotlight", null)
    }

    private val currentText: String
        get() = ttsEditText.text?.toString().orEmpty()

    private fun setText(text: String) {
        ttsEditText.setText(text)
        ttsEditText.setSelection(ttsEditText.text?.length ?: 0)
        updateActionButtonsState()
    }

    private fun updateActionButtonsState() {
        val hasText = currentText.isNotBlank()
        sayButton.isEnabled = isSpeaking || hasText
    }

    private fun observeTtsEvents() {
        val lifecycleOwner = findViewTreeLifecycleOwner()
        if (lifecycleOwner == null) {
            post { observeTtsEvents() }
            return
        }
        eventsJob?.cancel()
        eventsJob = lifecycleOwner.lifecycleScope.launch {
            lifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                tts.events().collect { event -> handleTtsEvent(event) }
            }
        }
    }

    private fun handleTtsEvent(event: TtsEvent) {
        when (event) {
            TtsEvent.SpeakingStarted -> {
                isSpeaking = true
            }

            TtsEvent.SpeakingCompleted -> {
                isSpeaking = false
                updateActionButtonsState()
            }

            is TtsEvent.Error -> {
                isSpeaking = false
                Toast.makeText(context, context.getString(R.string.tts_status_error, event.message), Toast.LENGTH_LONG).show()
                updateActionButtonsState()
            }

            is TtsEvent.Status -> Unit

            TtsEvent.DownloadStarted -> Unit

            is TtsEvent.DownloadProgress -> Unit

            is TtsEvent.DownloadCompleted -> {
                isSpeaking = false
                updateActionButtonsState()
            }
        }
    }

    private open class SimpleTextWatcher : android.text.TextWatcher {
        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) = Unit
        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) = Unit
        override fun afterTextChanged(s: android.text.Editable?) = Unit
    }
}
